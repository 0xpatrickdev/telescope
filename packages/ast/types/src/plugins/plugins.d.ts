export { getPluginValue } from '@cosmology/utils';
