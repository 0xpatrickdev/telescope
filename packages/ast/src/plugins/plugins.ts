import { TelescopeOptions, TelescopeOption } from '@osmonauts/types';

const getAllPackageParts = (name: string, list?: string[]) => {
    if (!list) list = [name];
    const newParts = name.split('.');
    newParts.pop();
    if (!newParts.length) return [...list];
    const newName = newParts.join('.');
    return getAllPackageParts(newName, [...list, newName]);
};

export const getPluginValue = (optionName: TelescopeOption, currentPkg: string, options: TelescopeOptions) => {
    const pkgOpts = options.packages;
    let value;
    getAllPackageParts(currentPkg).some((pkg, i) => {
        if (pkgOpts.hasOwnProperty(pkg)) {
            if (pkgOpts[pkg].hasOwnProperty(optionName)) {
                value = pkgOpts[pkg][optionName];
                return true;
            }
        }
    });

    if (value === undefined) {
        const defaultValue = options.hasOwnProperty(optionName) ? options[optionName] : undefined;
        value = defaultValue;
    }
    return value;
};
