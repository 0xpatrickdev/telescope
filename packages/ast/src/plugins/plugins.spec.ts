import cases from 'jest-in-case';
import { getPluginValue } from './plugins';

const options = {
    includeAminos: false,
    includeLCDClients: false,
    useDate: false,

    aminoEncoding: {
        enabled: true
    },

    packages: {
        cosmos: {
            useDate: true
        },
        'cosmos.v1beta1.tx': {
            includeAminos: false,
            useDate: false
        },
        'cosmos.v1beta1': {
            includeAminos: true,
        },
        osmosis: {
            aminoCasing: 'camel'
        },
        akash: {
            aminoEncoding: {
                enabled: false
            }
        },
        nebula: {
            some: {
                pkg: {
                    aminoEncoding: {
                        enabled: false
                    }
                }
            }
        }
    }
};


cases('useDate', opts => {
    const val = getPluginValue(opts.prop, opts.name, options);
    expect(val).toMatchSnapshot();

}, [
    { name: 'cosmos.v1beta1.tx', prop: 'useDate' },
    { name: 'cosmos.v1beta1', prop: 'useDate' },
    { name: 'cosmos.v1beta1.other', prop: 'useDate' },
    { name: 'cosmos', prop: 'useDate' },
]);

cases('aminoEncoding', opts => {
    const val = getPluginValue(opts.prop, opts.name, options);
    expect(val).toMatchSnapshot();

}, [
    { name: 'cosmos.v1beta1', prop: 'aminoEncoding' },
    { name: 'cosmos', prop: 'aminoEncoding' },
    { name: 'akash', prop: 'aminoEncoding' },
    { name: 'nebula', prop: 'aminoEncoding' },
    { name: 'nebula.some', prop: 'aminoEncoding' },
    { name: 'nebula.some.pkg', prop: 'aminoEncoding' },
]);

cases('aminoEncoding.enabled', opts => {
    const val = getPluginValue(opts.prop, opts.name, options);
    expect(val).toMatchSnapshot();

}, [
    { name: 'cosmos.v1beta1', prop: 'aminoEncoding.enabled' },
    { name: 'cosmos', prop: 'aminoEncoding.enabled' },
    { name: 'akash', prop: 'aminoEncoding.enabled' },
    { name: 'nebula', prop: 'aminoEncoding.enabled' },
    { name: 'nebula.some', prop: 'aminoEncoding.enabled' },
    { name: 'nebula.some.pkg', prop: 'aminoEncoding.enabled' },
]);