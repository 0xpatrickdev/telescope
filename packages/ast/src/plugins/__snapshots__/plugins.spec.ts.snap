// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`includeAminos cosmos 1`] = `false`;

exports[`includeAminos cosmos.v1beta1 1`] = `true`;

exports[`includeAminos cosmos.v1beta1.other 1`] = `true`;

exports[`includeAminos cosmos.v1beta1.tx 1`] = `false`;

exports[`useDate cosmos 1`] = `true`;

exports[`useDate cosmos.v1beta1 1`] = `true`;

exports[`useDate cosmos.v1beta1.other 1`] = `true`;

exports[`useDate cosmos.v1beta1.tx 1`] = `false`;
